### PR TITLE
BS-14567 Add method in PageAPI to retrieve a Page from a mapping key

### PR DESCRIPTION
**Pull request by rbioteau**
We need this method to avoid creating a mapping with a key that already
exists.

### CODE REVIEW

File: bpm/bonita-api/bonita-common-api/src/main/java/org/bonitasoft/engine/api/PageAPI.java [hunk 1]
```diff
@@ -220,5 +220,16 @@ Properties getPageProperties(byte[] content, boolean checkIfItAlreadyExists) thr
     PageURL resolvePageOrURL(String key, Map<String, Serializable> context, boolean executeAuthorizationRules) throws NotFoundException, UnauthorizedAccessException, ExecutionException;
 
 
+    /**
+     * Retrieves a page from its pathTemplate.
```
Comment (baptistemesta) [inline bpm/bonita-api/bonita-common-api/src/main/java/org/bonitasoft/engine/api/PageAPI.java]: path template don't really exists in the engine

File: bpm/bonita-core/bonita-process-engine/src/main/java/org/bonitasoft/engine/api/impl/page/PageAPIDelegate.java [hunk 2]
```diff
@@ -220,6 +220,17 @@ public Page getPageByName(final String name) throws PageNotFoundException {
         }
     }
 
+    public Page getPageByMappingKey(String mappingKey) throws PageNotFoundException {
+        try {
+            final SPageMapping sPageMapping = pageMappingService.get(mappingKey);
+            return convertToPage(pageService.getPage(sPageMapping.getId()));
+        } catch (final SBonitaReadException e) {
```
Comment (baptistemesta) [inline bpm/bonita-core/bonita-process-engine/src/main/java/org/bonitasoft/engine/api/impl/page/PageAPIDelegate.java]: SBonitaRead != page not found

Use RetrieveException which is runtime

File: bpm/bonita-core/bonita-process-engine/src/test/java/org/bonitasoft/engine/api/impl/page/PageAPIDelegateTest.java [hunk 3]
```diff
@@ -459,4 +461,40 @@ public void should_getPageProperties_throw_Invalid_token() throws Exception {
...
+    public void should_throw_a_PageNotFoundException_when_no_page_is_found_for_a_given_mapping_key() throws Exception {
+        // given
+        when(pageMappingService.get("apiExtension|GET|myPathTemplate")).thenThrow(SObjectNotFoundException.class);
+        // when
+        expectedException.expect(PageNotFoundException.class);
+        pageAPIDelegate.getPageByMappingKey("apiExtension|GET|myPathTemplate");
+    }
+
+    @Test
+    public void should_throw_a_PageNotFoundException_when_no_page_is_found_for_id_in_mapping() throws Exception {
```
Comment (baptistemesta) [inline bpm/bonita-core/bonita-process-engine/src/test/java/org/bonitasoft/engine/api/impl/page/PageAPIDelegateTest.java]: not not found but retrieve exception

**Comment by baptistemesta** [comment]
appart for the SBonitaReadException that should be converted to a RetrieveException, everything looks good.

File: bpm/bonita-core/bonita-process-engine/src/main/java/org/bonitasoft/engine/api/impl/page/PageAPIDelegate.java [hunk 4]
```diff
@@ -220,6 +220,17 @@ public Page getPageByName(final String name) throws PageNotFoundException {
...
+            throw new PageNotFoundException(e);
+        } catch (final SObjectNotFoundException e) {
+            throw new RetrieveException(String.format("Failed to retrieve a page mapping for %s", mappingKey));
```
Comment (baptistemesta) [inline bpm/bonita-core/bonita-process-engine/src/main/java/org/bonitasoft/engine/api/impl/page/PageAPIDelegate.java]: That's the other way around :)

File: bpm/bonita-core/bonita-process-engine/src/test/java/org/bonitasoft/engine/api/impl/page/PageAPIDelegateTest.java [hunk 5]
```diff
@@ -459,4 +461,40 @@ public void should_getPageProperties_throw_Invalid_token() throws Exception {
...
+        // given
+        final SPageMapping mapping = mock(SPageMapping.class);
+        final SPage page = mock(SPage.class);
+        when(mapping.getId()).thenReturn(1L);
+        when(pageService.getPage(1L)).thenThrow(SBonitaReadException.class);
```
Comment (baptistemesta) [inline bpm/bonita-core/bonita-process-engine/src/test/java/org/bonitasoft/engine/api/impl/page/PageAPIDelegateTest.java]: thenThrow SObjectNotFoundException

File: bpm/bonita-core/bonita-process-engine/src/test/java/org/bonitasoft/engine/api/impl/page/PageAPIDelegateTest.java [hunk 6]
```diff
@@ -459,4 +461,40 @@ public void should_getPageProperties_throw_Invalid_token() throws Exception {
...
+        when(pageMappingService.get("apiExtension|GET|myPathTemplate")).thenReturn(mapping);
+        // when
+        expectedException.expect(PageNotFoundException.class);
+        pageAPIDelegate.getPageByMappingKey("apiExtension|GET|myPathTemplate");
+    }
 }
```
Comment (baptistemesta) [inline bpm/bonita-core/bonita-process-engine/src/test/java/org/bonitasoft/engine/api/impl/page/PageAPIDelegateTest.java]: add test on SBonitaReadException (not mandatory but to be fully tested)

Comment (rbioteau) [inline bpm/bonita-core/bonita-process-engine/src/test/java/org/bonitasoft/engine/api/impl/page/PageAPIDelegateTest.java]: Updated

**Comment by baptistemesta** [comment]
good

**Pull request merged**